add cuopt direct solver

Iroy30 opened the pull request:
<!-- ##################################################################### -->
<!-- PLEASE READ BEFORE OPENING THIS PULL REQUEST -->

<!-- All changes must adhere to PEP8 standards as enforced by Black. -->
<!-- If your changes do NOT adhere, the test suite will fail. -->
<!-- Please read our Contributing guide for instructions on how to apply these standards. -->
<!-- Contributing Guide: https://pyomo.readthedocs.io/en/stable/contribution_guide.html -->
<!-- ##################################################################### -->

<!-- DO NOT DELETE OR IGNORE THIS TEMPLATE. Failing to adhere to this template and provide the necessary information may lead to your Pull Request being closed without consideration. -->

## Fixes #3626 

## Summary/Motivation:
Add cuOpt math optimization (includes LP and MILP) solver backend to Pyomo so users can solve pyomo models with cuOpt

## Changes proposed in this PR:
- Add cuopt_direct solver plugin 

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

Iroy30 committed:
add cuopt direct solver

mrmundt reviewed (commented):
We are working on making cuopt available in our testing infrastructure; can you please add tests to this PR?

Iroy30 committed:
add tests with lp amd milp capabilities

Iroy30 committed:
Merge remote-tracking branch 'origin/main'

Iroy30 commented:
@mrmundt Thanks! We have added tests by enabling testing cuopt with LP and MILP capabilities in tests/solvers.py. Let us know if:

1. There is any other testing that needs to be enabled. 
2. For testing cuOpt backend in CI, it needs a GPU, is that something pyomo team intends to add or will it be disabled in github CI?

The following is the testing output I get relevant to cuOpt
```
Testing scenario generation
------------------------------
LP_block, cuopt, python
   ok: 
LP_block_kernel, cuopt, python
   ok: 
LP_compiled, cuopt, python
   ok: 
LP_compiled_dense_kernel, cuopt, python
   ok: 
LP_compiled_sparse_kernel, cuopt, python
   ok: 
LP_constant_objective1, cuopt, python
   ok: 
LP_constant_objective1_kernel, cuopt, python
   ok: 
LP_constant_objective2, cuopt, python
   ok: 
LP_constant_objective2_kernel, cuopt, python
   ok: 
LP_duals_maximize, cuopt, python
   ok: 
LP_duals_maximize_kernel, cuopt, python
   ok: 
LP_duals_minimize, cuopt, python
   ok: 
LP_duals_minimize_kernel, cuopt, python
   ok: 
LP_inactive_index, cuopt, python
   ok: 
LP_inactive_index_kernel, cuopt, python
   ok: 
LP_infeasible1, cuopt, python
   ok: 
LP_infeasible1_kernel, cuopt, python
   ok: 
LP_infeasible2, cuopt, python
   ok: 
LP_infeasible2_kernel, cuopt, python
   ok: 
LP_piecewise, cuopt, python
   ok: 
LP_piecewise_nosuffixes, cuopt, python
   ok: 
LP_simple, cuopt, python
   ok: 
LP_simple_kernel, cuopt, python
   ok: 
LP_trivial_constraints, cuopt, python
   ok: 
LP_trivial_constraints_kernel, cuopt, python
   ok: 
LP_unbounded, cuopt, python
   ok: 
LP_unbounded_kernel, cuopt, python
   ok: 
LP_unused_vars, cuopt, python
   ok: 
LP_unused_vars_kernel, cuopt, python
   ok: 
MILP_discrete_var_bounds, cuopt, python
   ok: 
MILP_discrete_var_bounds_kernel, cuopt, python
   ok: 
MILP_infeasible1, cuopt, python
   ok: 
MILP_infeasible1_kernel, cuopt, python
   ok: 
MILP_simple, cuopt, python
   ok: 
MILP_simple_kernel, cuopt, python
   ok: 
MILP_unbounded, cuopt, python
   ok: 
MILP_unbounded_kernel, cuopt, python
   ok: 
MILP_unused_vars, cuopt, python
   ok: 
MILP_unused_vars_kernel, cuopt, python
   ok: 

```

mrmundt commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,261 @@
+            repn = generate_standard_repn(con.body, quadratic=False)
+            matrix_data.extend(repn.linear_coefs)
+            matrix_indices.extend([self.var_name_dict[str(i)] for i in repn.linear_vars])
+            """for v, c in zip(con.body.linear_vars, con.body.linear_coefs):
+                matrix_data.append(value(c))
+                matrix_indices.append(self.var_name_dict[str(v)])"""
```
Leftover debugging? Or intended to be included?

mrmundt commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,261 @@
+        log_file = None
+        if self._log_file:
+            log_file = self._log_file
+        t0 = time.time()
+        self.solution = cuopt.linear_programming.solver.Solve(self._solver_model)
+        t1 = time.time()
```
This is fine, but just so you're aware, we have this lovely little utility called `TicTocTimer` that you may want to consider using: https://pyomo.readthedocs.io/en/latest/api/pyomo.common.timing.TicTocTimer.html

whart222 commented:
@mrmundt What do you think about including this solver interface in pyomo.contrib.solvers?  Would it make sense to pull-in new solver interfaces there, since that's where the new solver API is evolving?

mrmundt commented:
> @mrmundt What do you think about including this solver interface in pyomo.contrib.solvers? Would it make sense to pull-in new solver interfaces there, since that's where the new solver API is evolving?

@whart222 - I am evenly split. Because we are still messing with what the new solver interfaces are going to actually do / how they will handle input and present output, I don't know if we want to put "new" solvers there or just "well-established" ones that we can robustly test / really know what they are supposed to do and return.

mrmundt commented:
@Iroy30 - Two more things:
1. Please run `black -S -C` on your files
2. Something in our testing environment isn't loading correctly. I'm looking into it.

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 committed:
Update pyomo/solvers/plugins/solvers/cuopt_direct.py

Co-authored-by: Miranda Mundt <55767766+mrmundt@users.noreply.github.com>

Iroy30 committed:
Merge remote-tracking branch 'origin'

Iroy30 committed:
formatting

Iroy30 committed:
pyomo/solvers/plugins/solvers/cuopt_direct.py

mrmundt committed:
Solver name incorrect - change to cuopt_direct

mrmundt commented:
Hi @Iroy30 - I got the tests running - the name in the `solvers.py` file was the culprit. That being said, there are quite a few errors (mentioned in the comment above).

It would be a good idea for you to create a `test_cuopt.py` file in `pyomo/solvers/tests/checks` and incorporate unit tests, small tests for basic functionality, little models that you know the answer for, etc.

Iroy30 commented:
I am looking at the "Attempting to use an unavailable solver" error currently. The rest should be fixed.

> Something in our testing environment isn't loading correctly. I'm looking into it.
Is this resolved?

Iroy30 commented:
we are investigating certain Termination condition mismatches on our end

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

mrmundt commented:
> @mrmundt Is there a way to disable certain tests temporarily for cuOpt solver as we improve on it? We are facing two CI failures:
> 
>     1. RC computation for a maximization problem failures
> 
>     2. cuOpt does not differentiate between infeasible vs unbounded at this point

Yeah, you can use `@unittest.skip` with an appropriate message. Are these real failures that need to be addressed, though?

Iroy30 committed:
add cuopt tests and fix CI fails

Iroy30 committed:
Merge remote-tracking branch 'origin/add_cuopt_direct_solver_plugin' into add_cuopt_direct_solver_plugin

Iroy30 commented:
There are active improvements and fixes in cuOpt that are being worked on for the next release. I have skipped two above mentioned tests temporarily. The CI tests are passing for me locally. Hopefully all runs successfully in the pipeline as well.

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

jsiirola commented:
cuOpt is not available on GHA because the GHA runners (to our knowledge) do not have GPUs.  The Jenkins job *will* run on machines with GPUs and cuOpt, and should exercise this interface.  Regardless, solver interfaces must be able to be imported and instantiated (to check for availability) without a hard dependency on any "external" packages.

A quick look at you implementation:
- you will need to guard the numpy import in `cuopt_direct.py` (the easy way it to import it from `pyomo.common.sdependencies`
- you should defer resolving the _version and checking if the python api exists until after (outside) the `CUOPTDirect.__init__()`.

Iroy30 committed:
update import

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 committed:
Merge remote-tracking branch 'origin/add_cuopt_direct_solver_plugin' into add_cuopt_direct_solver_plugin

Iroy30 commented:
@jsiirola Thanks John, aded the guards and updated version, hopefully this should fix the fails when cuopt isn't present. Should we re-run CI?

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 commented:
@mrmundt @jsiirola How can I go ahead and re-run the CI?

Iroy30 commented:
We are gearing up for launch of cuopt 25.10, it would be great if we could review this PR before the new release so that we are in sync

mrmundt reviewed (changes requested):
Thanks, @Iroy30 , for this PR! I have done an initial review. Some high-level things to consider:
1. Logging - `logger` is created but then never used. Consider places it may be helpful
2. "Magic" numbers - there are quite a few instances of numbers that aren't defined being used. We love documentation!

mrmundt committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

jsiirola reviewed (changes requested):
This is actually pretty close, but there are a number of things that need to be resolved before we can merge.

Iroy30 commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,360 @@
+from pyomo.solvers.plugins.solvers.direct_or_persistent_solver import (
+    DirectOrPersistentSolver,
+)
+from pyomo.core.kernel.objective import minimize, maximize
```
updated

Iroy30 commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,360 @@
+
+        prob_type = solution.problem_category
+
+        if status in [1]:
+            self.results.solver.status = SolverStatus.ok
+            self.results.solver.termination_condition = TerminationCondition.optimal
+            soln.status = SolutionStatus.optimal
+        elif status in [3]:
+            self.results.solver.status = SolverStatus.warning
+            self.results.solver.termination_condition = TerminationCondition.unbounded
+            soln.status = SolutionStatus.unbounded
+        elif status in [8]:
+            self.results.solver.status = SolverStatus.ok
+            self.results.solver.termination_condition = TerminationCondition.feasible
+            soln.status = SolutionStatus.feasible
+        elif status in [2]:
+            self.results.solver.status = SolverStatus.warning
+            self.results.solver.termination_condition = TerminationCondition.infeasible
+            soln.status = SolutionStatus.infeasible
+        elif status in [4]:
+            self.results.solver.status = SolverStatus.aborted
+            self.results.solver.termination_condition = (
+                TerminationCondition.maxIterations
+            )
+            soln.status = SolutionStatus.stoppedByLimit
+        elif status in [5]:
+            self.results.solver.status = SolverStatus.aborted
+            self.results.solver.termination_condition = (
+                TerminationCondition.maxTimeLimit
+            )
+            soln.status = SolutionStatus.stoppedByLimit
+        elif status in [7]:
```
Added Comment

Iroy30 commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,360 @@
+from pyomo.opt.results.solution import Solution, SolutionStatus
+from pyomo.opt.results.solver import TerminationCondition, SolverStatus
+from pyomo.opt.base import SolverFactory
+from pyomo.core.base.suffix import Suffix
```
removed

Iroy30 commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,360 @@
+
+            conname = self._symbol_map.getSymbol(con, self._labeler)
+            self._pyomo_con_to_solver_con_map[con] = con_idx
+            con_idx += 0
```
updated. Good catch. We use it when storing dual solution

Iroy30 commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,360 @@
+            repn = generate_standard_repn(con.body, quadratic=False)
+            matrix_data.extend(repn.linear_coefs)
+            matrix_indices.extend(
+                [self.var_name_dict[str(i)] for i in repn.linear_vars]
```
updated

Iroy30 commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,360 @@
+        self._solver_model.set_constraint_lower_bounds(np.array(c_lb))
+        self._solver_model.set_constraint_upper_bounds(np.array(c_ub))
+
+    def _add_var(self, variables):
```
updated

Iroy30 commented on pyomo/solvers/tests/checks/test_cuopt_direct.py:
```diff
@@ -0,0 +1,82 @@
+from pyomo.common.tempfiles import TempfileManager
+import pyomo.common.unittest as unittest
+
+try:
+    import cuopt
+
+    cuopt_available = True
+except:
+    cuopt_available = False
```
has been updated to attempt_import

Iroy30 commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,360 @@
+
+        prob_type = solution.problem_category
+
+        if status in [1]:
```
updated

Iroy30 committed:
address review comments part 1

mrmundt committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 commented:
@mrmundt Would it be possible to update CI to pull cuopt 25.10

Iroy30 commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,437 @@
+        try:
+            self._solver_model = cuopt.linear_programming.DataModel()
+        except Exception:
+            e = sys.exc_info()[1]
```
resolved

Iroy30 commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,360 @@
+                    raise ValueError(
+                        "Solver interface does not support multiple objectives."
+                    )
+                self._set_objective(obj)
```
updated

Iroy30 commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,360 @@
+        extract_reduced_costs = False
+        for suffix in self._suffixes:
+            flag = False
+            if re.match(suffix, "dual"):
```
leaving as is, marking resolved

Iroy30 committed:
address review comments part 2

Iroy30 committed:
Merge remote-tracking branch 'iroy30/add_cuopt_direct_solver_plugin'

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,437 @@
+
+cuopt, cuopt_available = attempt_import("cuopt")
+
+def toDict(model, json=False):
```
removed the helper/debug function

Iroy30 committed:
remove print

Iroy30 commented:
@jsiirola The segfaults you see in CI with 25.08 is concerning. Was the CI running the 25.08 release or the daily build of cuopt(this could cause the segfault). If 25.08 release was causing segfault then we might have to enforce minimum cuopt 25.10 and meanwhile I will debug 25.08 on my end because we were not seeing this previously

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

bbozkaya commented:
@mrmundt, @jsiirola would it help communicate and resolve faster if we had a Slack channel set up for communication?
If yes, please provide an email address to which we can send a Slack join request.

Iroy30 commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,360 @@
+        self.results.problem.lower_bound = None
+        try:
+            self.results.problem.upper_bound = solution.get_primal_objective()
+            self.results.problem.lower_bound = solution.get_primal_objective()
```
Updated. We have the obj bound populated from milp stats

jsiirola committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 committed:
address review part 3, black formatting

Iroy30 committed:
Merge remote-tracking branch 'iroy30/add_cuopt_direct_solver_plugin'

Iroy30 commented:
@jsiirola Did the instructions fix the cuOpt import?

mrmundt committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

mrmundt commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,371 @@
+    def __init__(self, **kwds):
+        kwds["type"] = "cuoptdirect"
+        super(CUOPTDirect, self).__init__(**kwds)
+        self._version = cuopt.__version__.split('.')
```
This is the problem line. If cuopt isn't actually available, then this panics. You should probably wrap this in a try block or move it into its own function, e.g.:
```
try:
    self._version = cuopt.__version__.split('.')
except DeferredImportError:
    self._version = None
```

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 committed:
move version resolution

Iroy30 committed:
reformat

jsiirola committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

jsiirola committed:
DeferredImportCallbackLoader needs to expose get_resource_reader()

jsiirola committed:
fix parsing of cuopt version

jsiirola committed:
Don't perform redundant attempt_import('cuopt')

Iroy30 committed:
delete instance level version

Iroy30 committed:
Update cuopt_direct.py

Iroy30 commented:
Great! Looks like the CI passes except [codecov/patch](https://app.codecov.io/gh/Pyomo/pyomo/pull/3620) , how do we reconcile this, any suggestions?

Iroy30 committed:
improve test coverage, add tests

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 committed:
Merge remote-tracking branch 'iroy30/add_cuopt_direct_solver_plugin'

blnicho committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

blnicho committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

blnicho committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

blnicho committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

jdwit committed:
Address review feedback

- Remove unused imports
- Use new-style super()
- Add logging (debug for solve, warnings for MIP limitations)

Iroy30 committed:
Merge pull request #1 from jdwit/fix/cuopt-improvements

Address requested changes

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,388 @@
+class CUOPTDirect(DirectSolver):
+    def __init__(self, **kwds):
+        kwds["type"] = "cuoptdirect"
+        super(CUOPTDirect, self).__init__(**kwds)
```
Updated

michaelbynum commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,360 @@
+        self._solver_model.set_variable_names(np.array(list(self.var_name_dict.keys())))
+
+    def _set_objective(self, objective):
+        repn = generate_standard_repn(objective.expr, quadratic=False)
```
@jsiirola - could you add a doc string to the `LinearRepnVisitor` that at least explains the `subexpression_cache` argument. Something that explains the `LinearRepn` (and its `multiplier` in particular) would probably also be helpful.

michaelbynum commented:
I think this is very close. I have one change I'd like to see.

jsiirola reviewed (commented):
Over all, this is in pretty good shape.  I have some comments, but the only thing that I think should be resolved before merging this is the use of `.name` as a "unique" identifier in the variable map.

jdwit committed:
use pyomo_var_to_ndx_map instead of var_name_dict for variable indexing

jdwit committed:
switch to LinearRepnVisitor for constraint processing

jdwit committed:
use LinearRepnVisitor for objective processing

jdwit committed:
minor cleanup and add test for trivial constraint

jdwit committed:
add nonlinear constraint check

Iroy30 committed:
Merge pull request #2 from jdwit/fix-var-name-dict

Address review feedback: use LinearRepnVisitor and fix var_name_dict

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 committed:
black formatting

Iroy30 committed:
black formatting

michaelbynum commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,416 @@
+
+    def _set_objective(self, objective):
+        visitor = LinearRepnVisitor({})
+        repn = visitor.walk_expression(objective.expr)
```
This also needs a check to ensure the expression is linear.

michaelbynum commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,416 @@
+        ----------
+        cons_to_load: list of Constraint
+        """
+        self._load_duals(cons_to_load)
```
Same here - is there a way to make sure the model is continuous?

michaelbynum commented on pyomo/common/dependencies.py:
```diff
@@ -486,6 +486,9 @@ def exec_module(self, module: ModuleType) -> None:
         return self._loader.load_module(fullname)
 
+    def get_resource_reader(self, fullname):
+        return self._loader.get_resource_reader(fullname)
```
@jsiirola - I have no idea what this is, so I'm relying on you to make sure this is fine.

Iroy30 committed:
address more reviews

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 committed:
reformatting

jsiirola commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -0,0 +1,431 @@
+                    const = repn.constant if repn.constant else 0
+                    lb_val = lb if lb is not None else -np.inf
+                    ub_val = ub if ub is not None else np.inf
+                    if not (lb_val <= const <= ub_val):
```
Should this test be using a tolerance?  What is cuOpt's default constraint feasibility tolerance?

jsiirola reviewed (approved):
I think this is OK to merge.  There's always more comments, but I don't see anything that I think needs to hold this up.

michaelbynum committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

michaelbynum commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -335,16 +340,18 @@ def _postsolve(self):
             soln_variables = soln.variable
             soln_constraints = soln.constraint
-            for i, pyomo_var in enumerate(var_map.keys()):
+            for pyomo_var in var_map.keys():
```
`for pyomo_var, i in var_map.items():`?

michaelbynum commented on pyomo/solvers/plugins/solvers/cuopt_direct.py:
```diff
@@ -335,16 +340,18 @@ def _postsolve(self):
+            for pyomo_var in var_map.keys():
                 if len(primal_solution) > 0 and pyomo_var in self.referenced_vars:
                     name = self._symbol_map.getSymbol(pyomo_var, self._labeler)
                     soln_variables[name] = {"Value": primal_solution[i]}
```
oh, you updated it for reduced costs. Just missed the primals.

Iroy30 committed:
Update cuopt_direct.py

Iroy30 committed:
Update pyomo/solvers/plugins/solvers/cuopt_direct.py

Co-authored-by: John Siirola <jsiirola@users.noreply.github.com>

michaelbynum reviewed (approved):
Once tests pass, I think this is good to go.

Iroy30 committed:
update rc and dual check, formatting

Iroy30 commented on pyomo/solvers/tests/checks/test_cuopt_direct.py:
```diff
@@ -0,0 +1,142 @@
+        m.x = Var(domain=NonNegativeReals)
+        m.obj = Objective(expr=m.x, sense=minimize)
+        # trivial constraint that is infeasible: 5 <= 3
+        m.bad_con = Constraint(expr=5 <= 3)
```
Thanks, is that the only error/fail you see?

michaelbynum replied:
No. Sorry, I got distracted. I think @jsiirola got you everything, though.

jsiirola commented:
We are seeing 3 errors:

On all versions of Python:
```
________________ CUOPTTests.test_infeasible_trivial_constraint _________________

self = <pyomo.solvers.tests.checks.test_cuopt_direct.CUOPTTests testMethod=test_infeasible_trivial_constraint>

    @unittest.skipIf(not cuopt_available, "The CuOpt solver is not available")
    def test_infeasible_trivial_constraint(self):
        m = ConcreteModel()
        m.x = Var(domain=NonNegativeReals)
        m.obj = Objective(expr=m.x, sense=minimize)
        # trivial constraint that is infeasible: 5 <= 3
>       m.bad_con = Constraint(expr=5 <= 3)
        ^^^^^^^^^

pyomo/solvers/tests/checks/test_cuopt_direct.py:121: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pyomo/core/base/block.py:573: in __setattr__
    self.add_component(name, val)
pyomo/core/base/block.py:1088: in add_component
    val.construct(data)
pyomo/core/base/disable_methods.py:119: in construct
    return base.construct(self, data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
pyomo/core/base/constraint.py:738: in construct
    self._setitem_when_not_present(index, rule(block, index))
pyomo/core/base/indexed_component.py:1144: in _setitem_when_not_present
    obj.set_value(value)
pyomo/core/base/constraint.py:934: in set_value
    return super().set_value(expr)
           ^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pyomo.core.base.constraint.ScalarConstraint object at 0x7f0ebeff31b0>
expr = False

    def set_value(self, expr):
        """Set the expression on this constraint."""
        # [ ... ]    
        elif expr.__class__ is bool:
>           raise ValueError(
                "Invalid constraint expression. The constraint "
                "expression resolved to a trivial Boolean (%s) "
                "instead of a Pyomo object. Please modify your "
                "rule to return Constraint.%s instead of %s."
                "\n\nError thrown for Constraint '%s'"
                % (expr, "Feasible" if expr else "Infeasible", expr, self.name)
            )
E           ValueError: Invalid constraint expression. The constraint expression resolved to a trivial Boolean (False) instead of a Pyomo object. Please modify your rule to return Constraint.Infeasible instead of False.
E           
E           Error thrown for Constraint 'bad_con'

pyomo/core/base/constraint.py:466: ValueError
------------------------------ Captured log call -------------------------------
ERROR    pyomo.core:constraint.py:741 Rule failed when generating expression for Constraint bad_con with index None:
ValueError: Invalid constraint expression. The constraint expression resolved to a trivial Boolean (False) instead of a Pyomo object. Please modify your rule to return Constraint.Infeasible instead of False.

Error thrown for Constraint 'bad_con'
ERROR    pyomo.core:block.py:1091 Constructing component 'bad_con' from data=None failed:
    ValueError: Invalid constraint expression. The constraint expression resolved to a trivial Boolean (False) instead of a Pyomo object. Please modify your rule to return Constraint.Infeasible instead of False.

Error thrown for Constraint 'bad_con'
```

Only on Python 3.13:
```
____________ Test_LP_piecewise.test_cuopt_python_nonsymbolic_labels ____________

self = <pyomo.solvers.tests.checks.test_writers.Test_LP_piecewise testMethod=test_cuopt_python_nonsymbolic_labels>

    def return_test(self):
>       return writer_test(self)
               ^^^^^^^^^^^^^^^^^

pyomo/solvers/tests/checks/test_writers.py:161: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pyomo/solvers/tests/checks/test_writers.py:115: in writer_test
    self.fail(
E   AssertionError: Solution mismatch for plugin test_cuopt_python_nonsymbolic_labels, python interface and problem type LP_piecewise
E   Difference in solution for x.value:
E   	Baseline - 0.0
E   	Current - 1.0
E   Gap: None
E   Status: optimal
E   Message: None
E   Objective:
E     obj:
E       Value: 1
E   Variable:
E     x:
E       Rc: 0
E       Value: 1
E     y:
E       Rc: 1
E       Value: 1
E   Constraint: No nonzero values
----------------------------- Captured stdout call -----------------------------
Setting parameter log_file to tmpsi7fbm7i.log
cuOpt version: 25.10.0, git hash: 99e549c, host arch: x86_64, device archs: 70-real,75-real,80-real,86-real,90a-real,100f-real,120a-real,120
CPU: Intel(R) Xeon(R) 
CUDA 12.9, device: NVIDIA T400 4GB
CUDA device UUID: 

Solving a problem with 2 constraints, 2 variables (0 integers), and 4 nonzeros
Problem scaling:
Objective coefficents range:          [1e+00, 1e+00]
Constraint matrix coefficients range: [5e-01, 1e+00]
Constraint rhs / bounds range:        [0e+00, 5e-01]
Variable bounds range:                [0e+00, 0e+00]

Third-party presolve is disabled, skipping
Objective offset 0.000000 scaling_factor 1.000000
Running concurrent

Dual simplex finished in 0.00 seconds, total time 0.00
Barrier finished in 0.01 seconds
   Iter    Primal Obj.      Dual Obj.    Gap        Primal Res.  Dual Res.   Time
      0 +0.00000000e+00 +0.00000000e+00  0.00e+00   7.07e-01     1.00e+00   0.197s
PDLP finished
Concurrent time:  0.198s, total time 0.200s
Solved with dual simplex
Status: Optimal   Objective: 1.00000000e+00  Iterations: 0  Time: 0.200s
```
and 
```
_____________ Test_LP_piecewise.test_cuopt_python_symbolic_labels ______________

self = <pyomo.solvers.tests.checks.test_writers.Test_LP_piecewise testMethod=test_cuopt_python_symbolic_labels>

    def return_test(self):
>       return writer_test(self)
               ^^^^^^^^^^^^^^^^^

pyomo/solvers/tests/checks/test_writers.py:161: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pyomo/solvers/tests/checks/test_writers.py:71: in writer_test
    model_class.post_solve_test_validation(self, results)
pyomo/solvers/tests/models/base.py:463: in post_solve_test_validation
    tester.assertEqual(
E   AssertionError: <TerminationCondition.infeasible: 'infeasible'> != <TerminationCondition.optimal: 'optimal'>
----------------------------- Captured stdout call -----------------------------
Setting parameter log_file to tmp5mxrn1ik.log
cuOpt version: 25.10.0, git hash: 99e549c, host arch: x86_64, device archs: 70-real,75-real,80-real,86-real,90a-real,100f-real,120a-real,120
CPU: Intel(R) Xeon(R) 
CUDA 12.9, device: NVIDIA T400 
CUDA device UUID: 

Solving a problem with 2 constraints, 2 variables (0 integers), and 4 nonzeros
Problem scaling:
Objective coefficents range:          [1e+00, 1e+00]
Constraint matrix coefficients range: [5e-01, 1e+00]
Constraint rhs / bounds range:        [0e+00, 5e-01]
Variable bounds range:                [0e+00, 0e+00]

Third-party presolve is disabled, skipping
Objective offset 0.000000 scaling_factor 1.000000
Running concurrent

Dual simplex finished in 0.00 seconds, total time 0.00
Barrier finished in 0.01 seconds
   Iter    Primal Obj.      Dual Obj.    Gap        Primal Res.  Dual Res.   Time
      0 +0.00000000e+00 +0.00000000e+00  0.00e+00   7.07e-01     1.00e+00   0.104s
PDLP finished
Dual Simplex Solve status Primal Infeasible
Concurrent time:  0.104s, total time 0.106s
Solved with dual simplex
Status: Primal Infeasible   Objective: nan  Iterations: 1  Time: 0.106s
```

Iroy30 committed:
fix trivial constraint test

Iroy30 committed:
Merge branch 'main' into add_cuopt_direct_solver_plugin

Iroy30 commented:
Ive fixed the first error. I am unable to reproduce the other two locally on my python 3.13 environment. Could be that is because I'm using the most recent cuOpt 26.02 version, I see that the CI runs on 25.10. I will have to downgrade the version and re-evaluate the fails

Iroy30 committed:
Update test_cuopt_direct.py

jsiirola committed:
NFC: apply black

blnicho commented:
@Iroy30 yes, this will be included in the 6.10 release which should be out in the next couple days
